Support @EnsuresNonNullIf

mauricioaniche opened the pull request:
This PR is an initial draft of a proposal to support `@EnsuresNonNullIf`, following the discussion on issue #1032. This annotation allows the definition of methods that ensure that fields aren't null. 

An example of the annotation:

```
class Foo {
  @Nullable Object field;

  @EnsuresNonNullIf("field")
  boolean hasField() {
    return field != null;
  }

  void action() {
    if(!hasField()) {
      return;
    }

    // field is non-null from this point on.
  }
}
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber/NullAway?pullRequest=1044) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/uber/NullAway?pullRequest=1044) before we can accept your contribution.<br/><hr/>**mauricio** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber/NullAway?pullRequest=1044) it.</sub>

mauricioaniche commented:
Force pushing just so I can change the user name in the commit, which wasn't matching my github user.

msridhar commented:
This looks very promising, thanks!  I'll just answer the questions for now and assume a full code review should wait for later.

> * For now, the validation mechanism expects a precise "return field != null;". Any deviation from that will trigger a NullAway error. Is that fine?

I think we can do a bit better without too much trouble.  Looking at the code [here](https://github.com/uber/NullAway/blob/f36ab25e77604b75907a7d834d8b3e3f3f7fd303/nullaway/src/main/java/com/uber/nullaway/handlers/StreamNullabilityPropagator.java#L603-L609), I think the `thenStore` and `elseStore` parameters should reflect the nullability info computed by the dataflow analysis when returning `true` and `false` respectively.  So, we should be able to look inside the appropriate store to see if the inferred type for the field is `@NonNull`.  So we should be able to handle cases like, e.g., `if (field != null) return true;`.

> * I still need to work on supporting multiple fields.

No problem.  I think if needed, a first version could even skip that as long as we track multiple field support as an issue for a follow up.

> * Should I support an extra parameter to the annotation, so the person can define whether the method check does == or != null?

Yes, I think that would be good for flexibility, and it shouldn't be too hard to implement.

> * During validation, I need to check whether the field exists in the class.

👍 It can exist in the class itself _or_ it can be inherited (see below).

> * What to do in `validateOverridingRules`?

We need to prevent cases like (rough syntax):
```java
class A {
  @Nullable Object foo;
  @EnsuresNonNullIf("this.foo")
  boolean hasFoo() { return this.foo != null; }
}
class B extends A {
  @Override
  boolean hasFoo() { return true; }
}
...
void test() {
  A a = new B();
  if (a.hasFoo()) {
    a.foo.toString(); // NPE!
  }
}
```
To prevent this, an overriding method should have at least as strong an `@EnsuresNonNullIf` annotation as the overridden method.  So if the superclass method ensures `f` and `g` are non-null when returning `true`, the subclass method should also ensure `f` and `g` are non-null when returning `true`, but it can also ensure `h` is non-null.

mauricioaniche commented:
@msridhar I added a few more features, like the post-condition check we spoke about and a `trueIfNonNull` boolean that inverts the condition.

The only thing missing is that if the if check is extracted to a boolean variable, the handler gets lost. See all tests that are disabled.

Thoughts?

mauricioaniche commented:
@msridhar Oh, that's good to know, I thought I was making a mistake somewhere! Then, I believe I coded everything I wanted for this feature.

* The new `EnsuresNonNullIf` accepts a list of fields and a boolean `trueIfNonNull` allowing the person to flip the outcome of the method.
* The handler takes care of semantics validation as well as post-conditions in the child classes
* A bunch of tests

Looking forward to your code review. More importantly, maybe you have some extra test cases you'd like me to try? I'm sure your eyes are way more trained than mine to spot corner cases!

msridhar commented:
@mauricioaniche will try to look at this one very soon.  In the meantime, are you able to sign the CLA?

mauricioaniche commented:
CLA signed!

mauricioaniche commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullIfHandler.java:
```diff
@@ -0,0 +1,257 @@
+    if (visitingAnnotatedMethod) {
+      // We might have already found another return tree that results in what we need,
+      // so we don't keep going deep.
+      if (semanticsHold) {
```
Any thoughts about this? A method with multiple returns, can we be smart about it?

msridhar replied:
See comment above about checking `return` statements individually

msridhar reviewed (commented):
Thanks again for the contribution!  It will take me some time to review this whole PR.  I am pushing my comments in chunks when I get time, to get the feedback out faster.

mauricioaniche commented on annotations/src/main/java/com/uber/nullaway/annotations/EnsuresNonNullIf.java:
```diff
@@ -0,0 +1,40 @@
+public @interface EnsuresNonNullIf {
+  String[] value();
+
+  boolean trueIfNonNull() default true;
```
I made it optional because my gut feeling says that the large majority of such methods are in the form of `hasX` rather than `doesNotHaveX`, which is more exceptional. I am more of a fan of sensible defaults instead of forcing people to explicit mark every decision, but of course, whatever you say here!

I don't get your comment about the "if" in the name. Should we change the name to something else? `trueIfNonNull` was just a placeholder as I couldn't think of something better!

msridhar replied:
> I made it optional because my gut feeling says that the large majority of such methods are in the form of `hasX` rather than `doesNotHaveX`, which is more exceptional. I am more of a fan of sensible defaults instead of forcing people to explicit mark every decision, but of course, whatever you say here!

For the `false` case I was thinking of methods like `StringUtils.isEmpty(s)` which ensures `s` is `@NonNull` if it returns `false`.  You can see a list of similar methods [here](https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java#L695-L727).  There might be analogous methods on fields.  I think you may be right that the other case is more common, but I'm not sure.

> I don't get your comment about the "if" in the name. Should we change the name to something else? `trueIfNonNull` was just a placeholder as I couldn't think of something better!

Sorry, I meant the `If` in `EnsuresNonNullIf`.  To me, given the annotation name, it is natural to want to specify the condition of the `If` part to be true or false.

Overall, I'd rather stick to doing something similar to the Checker Framework here.  If we really want to simplify the defaults, I'd rather have a separate annotation like `@EnsuresNonNullIfTrue`; but I feel that it's overkill.

mauricioaniche commented on nullaway/src/main/java/com/uber/nullaway/handlers/AbstractFieldContractHandler.java:
```diff
@@ -53,6 +53,11 @@ public abstract class AbstractFieldContractHandler extends BaseNoOpHandler {
   /** Simple name of the annotation in {@code String} */
   protected final String annotName;
 
+  // Set to true in case we are visiting a method with the annotation under analysis
+  protected boolean visitingAnnotatedMethod;
+  // Points to the method symbol that's currently being visited
+  protected Symbol.MethodSymbol visitingMethodSymbol;
+
```
Makes sense to me! I indeed saw the use of the map to keep the state there and I was wondering why, didn't think of anonymous classes inside a method. I'll work on it.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullIfHandler.java:
```diff
@@ -0,0 +1,257 @@
+  protected boolean validateAnnotationSemantics(
+      MethodTree tree, MethodAnalysisContext methodAnalysisContext) {
+    // If no body in the method, return false right away
+    // TODO: Do we need proper error message here?
```
Would this happen for an abstract / interface method?  If so, seems best to report some kind of error?

mauricioaniche replied:
The existing `EnsuresNonNullHandler` actually returns `true` in this case. I'm doing the same now, as someone probably thought deeper about it!

mauricioaniche replied:
Keeping this one open for you to close it yourself, if you agree!

msridhar commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullIfHandler.java:
```diff
@@ -0,0 +1,257 @@
+    VisitorState state = methodAnalysisContext.state();
+    analysis
+        .getNullnessAnalysis(state)
+        .forceRunOnMethod(new TreePath(state.getPath(), tree), state.context);
```
See my other comment where I indicated what might be a good place to call `forceRunOnMethod`

msridhar commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullIfHandler.java:
```diff
@@ -0,0 +1,257 @@
+        .forceRunOnMethod(new TreePath(state.getPath(), tree), state.context);
+
+    // If listed fields aren't checked for their nullness, return error
+    if (!semanticsHold) {
```
I think reporting errors here again won't quite work in the presence of nested methods.  Instead, I suggest keeping another map from method trees to the information from the annotation (the field names and whether they should be non-null when returning `true` or `false`).  Then, we should separately check each `return` statement in `onDataflowVisitReturn` and report an error directly at that point.  (This will also give better error locations.)  Right now the `VisitorState` (needed to call `reportMatch`) is not available as a parameter in `onDataflowVisitReturn`, but it should be easy to pass in if you add it as an additional parameter (it's available in a field [at this call site](https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessPropagation.java#L855-L855)).

mauricioaniche replied:
Oh I only saw this comment now. Will try to move the validation to each call to the onDataflowVisitReturn. 

Not sure I will need the map, as I believe I already have everything I need there, besides the VisitorState, which I'll try to pass it on.

msridhar reviewed (commented):
A few more comments.  I'll do another full pass after they've been addressed.  Thanks again!

msridhar commented on nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java:
```diff
@@ -160,6 +160,7 @@ public Set<Element> getNonnullFieldsOfReceiverAtExit(TreePath path, Context cont
     return getNonnullReceiverFields(nullnessResult);
   }
 
+  // TODO re-use this code
```
I added this TODO.  In this PR it'd be good if we could find a way to re-use this code, which finds all the `@NonNull` fields in a store

mauricioaniche commented:
@msridhar I am working on moving the "visitingMethod" state from the abstract class to the class itself, just so we can better support code that may use anonymous classes. It isn't being that straightforward. I mostly had to keep maps that convert ErrorProne's data structure to CheckerFramework's and vice-versa. It feels like I always need `MethodAnalysisContext` within ErrorProne's AST visitor and `Tree`s in CheckerFramework dataflow callbacks 😆 

My issue right now is that I can't use the `validateAnnotationSemantics`, as I need the data-flow engine (and ErrorProne's AST visitor) to fully run before I can decide if the semantics are correct. Before, I would call the data-flow engine from within this validate method, and then I'd have all the information I needed to validate. However, now that I have to visit each return tree (like the stream propagation logic), I can only take this decision after I visited all the return trees, which happens only after the provided `validateAnnotationSemantics` callback. ErrorProne doesn't give us a "end visit" callback that I could use. 

Right now I'm doing it on `onMatchMethodInvocation`, which means validation happens now the first time the method is called. But now that I'm writing this message, I realize that my code will fail if the declaration of the EnsuresNonNullIf method happens after the method invocation. Any thoughts on how I can do this semantic validation better?

msridhar commented:
@mauricioaniche I think for this annotation, it doesn't make sense to do all the validation inside `validateAnnotationSemantics`.  Perhaps we can rename that method or make it more specific to the `EnsuresNonNullHandler` as I think that's the only handler that does actual validation there.  Instead, for this handler, I think we can force the dataflow to run in `onMatchReturn` as suggested previously.  That will force the calls to `onDataflowVisitReturn` where we will validate each `return` statement separately, passing in a `VisitorState` object so we can report errors directly.  Does that make sense?

mauricioaniche commented:
@msridhar Im currently calling the data-flow in the onMatchReturn which does trigger the data-flow for all returns, and onDataflowVisitReturn is nicely called for every return statement. The problem is that, at that point, I don't have the complete map of return statements and their enclosing methods (which has to be built via the onMatchReturn callback). If there's a return inside an anonymous class, I can't identify that. 

I didn't see a way to get or pass the enclosing method in the onDataflowVisitReturn.

Maybe there's some other onDataFlow callback I can listen to to understand I'm leaving the scope of the current method. I'll check that today.

msridhar commented:
@mauricioaniche you are right, I didn't think of that.  I'll try to think of an alternative solution.

msridhar commented:
@mauricioaniche not 100% sure of all the details but I think we may need to do something like run our own visitor traversal inside `onMatchMethod` (after forcing the dataflow analysis to run) rather than trying to use `onDataflowVisitReturn`.  There is code like this where we try to verify contracts:

https://github.com/uber/NullAway/blob/940c40edda0e972df7fcb3543f514a14de661917/nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java#L101-L101

Perhaps something like this could work?  The thing is I think even this code doesn't handle nested methods correctly...but maybe we are getting closer.  Will think more, sorry for all this trickiness.

mauricioaniche commented:
I'll read the code you sent me for inspiration!

Another alternative could be to not support nested methods in the EnsuresNonNullIf annotated method. I would not expect too much complexity in such methods anyways. Let me also investigate if we could detect such "violation".

mauricioaniche commented on nullaway/src/main/java/com/uber/nullaway/NullabilityUtil.java:
```diff
@@ -255,6 +248,22 @@ public static Stream<? extends AnnotationMirror> getAllAnnotations(Symbol symbol
     return null;
   }
 
+  public static @Nullable AnnotationMirror findAnnotation(
```
Done

mauricioaniche commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/FieldContractUtils.java:
```diff
@@ -0,0 +1,70 @@
+      if (iterator.hasNext()) {
+        errorMessage.append(", ");
+      }
+    }
```
Done

mauricioaniche commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullIfHandler.java:
```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (c) 2017-2020 Uber Technologies, Inc.
```
Done

mauricioaniche commented:
@msridhar I believe I made it work. The code is now similar to my previous version, so much simpler. The trick:

In the `validateAnnotationSemantics`, build a set of `ReturnTree`s that are directly inside the method under validation. This is possible due to a on-the-spot `TreePathScanner` visit. Then, I call the data-flow engine. The `onDataFlowVisitReturn` does the same validation logic as before, but now it skips return trees that are not in the pre-built set. Once back to `validateAnnotationSemantics`, I can check the semantics held; if not, raise an error.

I added new tests to ensure that the order of method declaration doesn't affect the result, and one to show that a return inside a lamba doesn't affect the process.

Let me know what you think!

mauricioaniche commented:
@msridhar I also applied the requested refactorings. Only thing missing is now to make the property of the annotation compulsory, which I'll do once you think the hard part is done!

mauricioaniche commented:
Only saw the message about doing the validation in each return, will try that! 

Although I believe it does work as is right now due to the fact that I only visit return trees that are in the method itself and not on nested ones, but it does seem more of a clean solution if I make it work!

msridhar commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullIfHandler.java:
```diff
@@ -0,0 +1,303 @@
+    new TreePathScanner<Void, Void>() {
+      @Override
+      public Void visitReturn(ReturnTree node, Void unused) {
+        VisitorState returnState = methodState.withPath(getCurrentPath());
+
+        TreePath enclosingMethod =
+            NullabilityUtil.findEnclosingMethodOrLambdaOrInitializer(returnState.getPath());
```
I don't think you need `returnState`?  Just do `NullabilityUtil.findEnclosingMethodOrLambdaOrInitializer(getCurrentPath())`?

mauricioaniche replied:
Yes!

msridhar commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullIfHandler.java:
```diff
@@ -0,0 +1,303 @@
+    return true;
+  }
+
+  private void buildUpReturnToEnclosingMethodMap(VisitorState methodState) {
```
This is a great approach!

msridhar commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullIfHandler.java:
```diff
@@ -0,0 +1,303 @@
+      return;
+    }
+
+    // We might have already found another return tree that results in
+    // what we need, so we don't keep going deep.
+    if (semanticsHold) {
+      return;
+    }
```
I don't quite follow this.  Don't we need the fields to be non-null at _every_ return statement under the right condition?  I read this logic as checking if the fields are non-null (the semantics hold) for at least one statement.

mauricioaniche replied:
Old comment, not in the current implementation, closing it!

msridhar commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullIfHandler.java:
```diff
@@ -0,0 +1,303 @@
+    // engine in the "then" case (i.e., true case)
+    // and check whether all fields in the annotation parameter are non-null
+    Set<String> nonNullFieldsInPath =
+        thenStore.getReceiverFields(trueIfNonNull ? Nullness.NONNULL : Nullness.NULL).stream()
```
I think this works, but it's a little non-obvious.  The dataflow analysis happens to track `Nullness.NULL` at this point, but I'm not sure I would rely on it.  Instead, I would check for `NONNULL` fields in either the `thenStore` or the `elseStore`, depending on `trueIfNonNull`.

msridhar replied:
I now think this logic isn't 100% correct, and that it's important to use `trueIfNonNull` to select the `thenStore` or the `elseStore`.  See comments in tests below.

msridhar commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullIfHandler.java:
```diff
@@ -0,0 +1,303 @@
+    if (allFieldsInPathAreVerified) {
+      // If it's a literal, then, it needs to return true/false,
+      // depending on the trueIfNonNull flag
+      if (returnTree.getExpression() instanceof LiteralTree) {
+        LiteralTree expressionAsLiteral = (LiteralTree) returnTree.getExpression();
+        if (expressionAsLiteral.getValue() instanceof Boolean) {
+          this.semanticsHold = (boolean) expressionAsLiteral.getValue();
+        }
```
I'm not sure I follow this.  I think from the comment above that if you choose the `thenStore` or `elseStore` and just check for `NONNULL` fields you shouldn't need this.  Those stores already correspond respectively to what is known when the return value is `true` or `false`, I think.

mauricioaniche replied:
You were indeed correct! The algorithm now picks the store based on the `result` flag!

msridhar reviewed (commented):
I think you have the nested methods problem solved!  Here are a few more comments

mauricioaniche commented:
@msridhar I have a yet another version! The semantics validation happen at each return statement. I wrote a comment describing how the decision is made as it involves the outcome of the data-flow engine and the (possible) literal boolean expression of the return tree.

Regarding the need for the "boolean literal evaluation", see for example, test `correctUse_moreComplexFlow`:

```
@EnsuresNonNullIf(value="nullableItem", result=true)
public boolean hasNullableItem() {
  if(nullableItem != null) {
    return true;
  } else {
    return false;
  }
}
``` 

When the data-flow engine is assessing the `return false` branch, there are zero non-null fields, both at the `then` and at the `else` stores. Correctly so. I sort of need to know that, if the method is returning false, this means (some) fields can still be nullable. In code, both `thenStore` and `elseStore` have this field as nullable, but I can't see how the code can understand that this is a "return false" so fields should be null without parsing the literal.

Any thoughts on how I can do that besides having to manually evaluate the literal boolean there?

msridhar commented:
> Any thoughts on how I can do that besides having to manually evaluate the literal boolean there?

You are completely right that you need logic like this.  Sorry I forgot about it.  What we want here is the following: assuming the annotation is for the case of returning `true` (the `false` case is analogous), we want the mentioned fields to be guaranteed to be non-null for any return statement that can _possibly_ return `true` (to be conservative).  The `StreamNullabilityPropagator` has simple logic for this that will assume the statement might return `true` unless it is literally `return false`:

https://github.com/uber/NullAway/blob/940c40edda0e972df7fcb3543f514a14de661917/nullaway/src/main/java/com/uber/nullaway/handlers/StreamNullabilityPropagator.java#L508-L522

Possibly we want to extract and share this logic, in case we decide to do some improved reasoning in the future.

msridhar commented:
FYI there is a CI failure due to [warnings](https://github.com/uber/NullAway/actions/runs/11091905257/job/30820851428?pr=1044#step:5:521) when running NullAway on itself.  You can repro locally by running `./gradlew :nullaway:buildWithNullAway`

mauricioaniche commented:
@msridhar I believe that's what I implemented in this latest version! Or am I missing something?

msridhar commented:
> Edit: in my case, if not a Boolean literal, I delegate to the knowledge of the data flow engine. In which cases can this fail?

Do you mean that if it's not a literal, you assume it might return either true or false and appropriately check if the fields are non-null according to the data flow engine?  If so, that is correct.

Sorry, I hadn't reviewed the code, only looked at your comment.  I'll do another pass over the code soon.

msridhar commented:
BTW, for quick questions feel free to [join the NullAway Discord](https://discord.gg/QH2F779) and ping me there.

mauricioaniche commented:
> Do you mean that if it's not a literal, you assume it might return either true or false and appropriately check if the fields are non-null according to the data flow engine? If so, that is correct.

Indeed. As you will see in the current code, if the return expression is a literal boolean, I then check whether or not "true" and "false" are the correct response given what I observe in the data-flow analysis. If the return expression is _not_ a literal boolean (meaning, maybe something more complex like an and expression), I then assume that, if the data-flow engine considered all fields as non-null in the "then store", it's then semantically correct. I wrote a bunch of tests for different possibilities, and they all seem green!

I'll join Discord!

mauricioaniche commented:
> FYI there is a CI failure due to [warnings](https://github.com/uber/NullAway/actions/runs/11091905257/job/30820851428?pr=1044#step:5:521) when running NullAway on itself. You can repro locally by running `./gradlew :nullaway:buildWithNullAway`

Thanks for the tip! It seems green now! 🟢

msridhar commented on nullaway/src/test/java/com/uber/nullaway/EnsuresNonNullIfTests.java:
```diff
@@ -0,0 +1,801 @@
+            "class Foo {",
+            "  @Nullable Item nullableItem;",
+            "  @Nullable Item nullableItem2;",
+            "  @EnsuresNonNullIf(value={\"nullableItem\",\"nullableItem2\"}, result=false)",
+            "  public boolean doesNotHaveNullableItem() {",
+            "    return nullableItem == null && nullableItem2 == null;",
+            "  }",
```
This annotation should _not_ pass validation.  If `nullableItem != null` but `nullableItem2 == null`, then this function returns `false`.  So returning `false` does not guarantee that both the fields are non-null.  I think using the `elseStore` (see comment above) will fix this problem.

mauricioaniche replied:
Your eyes are super trained for this!!! Great catch!

msridhar commented on nullaway/src/test/java/com/uber/nullaway/EnsuresNonNullIfTests.java:
```diff
@@ -0,0 +1,801 @@
+            "class Foo {",
+            "  @Nullable Item nullableItem;",
+            "  @Nullable Item nullableItem2;",
+            "  @EnsuresNonNullIf(value={\"nullableItem\",\"nullableItem2\"}, result=false)",
+            "  public boolean doesNotHaveNullableItem() {",
+            "    // BUG: Diagnostic contains: Method is annotated with @EnsuresNonNullIf but does not ensure fields [nullableItem, nullableItem2]",
+            "    return nullableItem == null || nullableItem2 == null;",
+            "  }",
```
This annotation _should_ pass validation.  If the function returns `false`, we know that neither of the fields can be `null`, i.e., both are non-null.

mauricioaniche replied:
Are you a person or a compiler? Great catch!

msridhar reviewed (commented):
Almost there, but there is a small logic bug that needs to be fixed I think

mauricioaniche commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullIfHandler.java:
```diff
@@ -0,0 +1,333 @@
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
```
done

mauricioaniche commented on nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullIfHandler.java:
```diff
@@ -0,0 +1,333 @@
+          throw new RuntimeException("no enclosing method, lambda or initializer!");
+        }
+
+        // We only add returns that are directly in the method under analysis
+        if (!enclosingMethod.getLeaf().equals(methodTreeUnderAnalysis)) {
+          return super.visitReturn(node, null);
+        }
+
+        // We add the current return tree to the set as it's in the method
+        returnTreesInMethodUnderAnalysis.add(node);
+        return super.visitReturn(node, null);
```
Good catch!

mauricioaniche commented on annotations/src/main/java/com/uber/nullaway/annotations/EnsuresNonNullIf.java:
```diff
@@ -0,0 +1,53 @@
+   *
+   * @return true or false
+   */
+  boolean result();
```
I'll then set the default of `result` to `true` as I truly believe this will simplify the work! I also think that `results` reads well!

mauricioaniche commented on annotations/src/main/java/com/uber/nullaway/annotations/EnsuresNonNullIf.java:
```diff
@@ -0,0 +1,53 @@
+   *
+   * @return The list of field names
+   */
+  String[] value();
```
I guess it's because CF accepts a true expression in that field. But I think `value` is the most common used name for the default/main setting of an annotation. I think it's fine as is!

mauricioaniche commented:
@msridhar you were right about the store! I was under the assumption that always looking at the "thenStore" was the right thing to do, because IIUC, it represents the path that follows the return expression. However, I didn't expect that the engine would not really be 100% accurate around the "NULL". Using the "elseStore" allowed me to look at NON_NULL always, which seems to be precise. Why does that happen? 

Moreover: I put it back the return default to true and fixed the minor nits as well.  Would you mind giving it a final check to see if everything looks fine?

Before merging: I think it's best to squash the commits just so the history of the repository doesn't get full of "bad commits". Do you want me to squash or do you do it via github?

msridhar commented:
@mauricioaniche thanks again for this amazing contribution! Will take one more look before merging. Don't worry we always squash and merge so the intermediate commits won't remain.

mauricioaniche commented:
@msridhar wait a sec, I believe I found a case where it breaks

mauricioaniche commented:
Take a look at the test `setResultToFalse_multipleElements_correctSemantics_2` that I just pushed, which is basically `setResultToFalse_multipleElements_correctSemantics_2` but with the `true` being returned as literal. This one should pass, right? It's failing now.

mauricioaniche commented:
@msridhar Done. It was a corner case on when return=false and the method returned a literal boolean. I had the feeling I had to also handle it. Main change is [here](https://github.com/uber/NullAway/pull/1044/commits/ea10969dee15839866950c2e4e6d880b79491c86#diff-909e79375ae97423b29b76587edf170a8cb7c2dc0e610046ae7b6708668287c5R206). I also renamed the variable and improved the documentation around it.

There was a wrong test case that was revealed after I fixed this bug! If you have the energy, please go through the test cases, and see if they are all correct!!

Thank you!

msridhar commented on nullaway/src/test/java/com/uber/nullaway/EnsuresNonNullIfTests.java:
```diff
@@ -0,0 +1,898 @@
+            "  @EnsuresNonNullIf(value=\"nullableItem\", result=true)",
+            "  public boolean hasNullableItem() {",
+            "    if(nullableItem != null) {",
+            "      // BUG: Diagnostic contains: The method ensures the non-nullability of the fields, but returns incorrectly",
```
I don't think this should be an error.  The annotation means, "If this method returns `true`, then `nullableItem` is guaranteed to be non-null."  If the method returns `false`, then the annotation says nothing about whether `nullableItem` will be null or not.  And at callers, we should only be using nullability information from the annotation for the case where the returned boolean matches the `result` field of the annotation.  So there is no need to perform any check within the method when we know it is not returning the `result` value.  The error reported for the `return true` statement is correct.  Do you agree?

mauricioaniche replied:
That does make sense to me! Let me work on it!

mauricioaniche replied:
Here: https://github.com/uber/NullAway/pull/1044/commits/56a0984a935cc89a0dd7819590afe6c7f4bd481e

msridhar commented on nullaway/src/test/java/com/uber/nullaway/EnsuresNonNullIfTests.java:
```diff
@@ -0,0 +1,898 @@
+            "  @EnsuresNonNullIf(value=\"nullableItem\", result=true)",
+            "  public boolean hasNullableItem() {",
+            "    if(nullableItem != null) {",
+            "      // BUG: Diagnostic contains: The method ensures the non-nullability of the fields, but returns incorrectly",
```
Same comment as the one below.

mauricioaniche replied:
Here https://github.com/uber/NullAway/pull/1044/commits/56a0984a935cc89a0dd7819590afe6c7f4bd481e

msridhar reviewed (commented):
Found another potential issue, see below

msridhar commented:
@mauricioaniche I think this ready to go!  I made a slight simplification to the store update logic in [fa7728e](https://github.com/uber/NullAway/pull/1044/commits/fa7728e140e0557b327cd86961be7f3a8d9ae37f), based on our updated understanding of how the "then" and "else" stores work.

Out of an abundance of caution, I'm going to run our benchmarking suite with this change, which involves pushing the branch up to our main repo and creating a dummy PR.  Assuming there are no red flags there I'll go ahead and squash and merge.

msridhar commented:
Hrm, it seems our benchmarking server needs a new JVM version so the benchmarking job is not working 😐 I'm not too worried about the perf impact of this change, so I will go ahead and land it.

mauricioaniche commented:
Great news! Thanks for your suppprt in this adventure, @msridhar !
